Add fedimint

TonyGiorgio opened the pull request:
TODO: 
- [x] payment storage in gluesql
- [x] Get labels to work properly with fedimint payments
- [x] NWC Invoice Payer

Follow up: 
- [ ] LNURL out of fedimint

benthecarman commented on mutiny-core/Cargo.toml:
```diff
@@ -41,6 +41,12 @@ aes = { version = "0.8" }
 
+fedimint-client = { git = "https://github.com/fedimint/fedimint", branch = "master" }
+fedimint-core = { git = "https://github.com/fedimint/fedimint", branch = "master" }
+fedimint-wallet-client = { git = "https://github.com/fedimint/fedimint", branch = "master"}
```
We might want to pick a commit and stick to it until 0.2.0, otherwise we might get random compile errors anytime they change an api

benthecarman commented on mutiny-core/src/fedimint.rs:
```diff
@@ -0,0 +1,139 @@
+        &DerivationPath::from(vec![
+            ChildNumber::from_hardened_idx(1)?,
+            ChildNumber::from_hardened_idx(fedimint_index.child_index)?,
+        ]),
```
we should incorporate the network into the secret, so mainnet is different than testnet

TonyGiorgio replied:
Shouldnt we really be doing that at the highest level instead? We should want this everywhere, not just for fedimint.

benthecarman replied:
We could but really wouldn't be following the BIPs anymore, normally for derivation paths it is `m/Purpose/Network/Account/Internal/Index`

TonyGiorgio replied:
Not going to be able to comply exactly with their derivations but the prefix essentially does once is goes into the fedimint client builder. I've also implemented network into this.

benthecarman commented on mutiny-core/src/fedimint.rs:
```diff
@@ -0,0 +1,139 @@
+pub struct FedimintIndex {
+    pub child_index: u32,
+    pub federation_code: String,
+    pub archived: Option<bool>,
```
dont think this needs to be optional, we only did that on node index for backwards compatibility

benthecarman commented on mutiny-core/src/nodemanager.rs:
```diff
@@ -708,6 +714,35 @@ impl<S: MutinyStorage> NodeManager<S> {
+            fedimint_map.insert(id, Arc::new(fedimint));
+        }
+
+        let fedimints = Arc::new(Mutex::new(fedimint_map));
```
should these be in lib instead of nodemanager? Imo that'd make more sense

TonyGiorgio replied:
moved, might still clean it up a bit in lib though.

benthecarman commented on mutiny-core/src/fedimint.rs:
```diff
@@ -0,0 +1,271 @@
+                    .await
+                    .map_err(|_| MutinyError::ConnectionFailed)?; // TODO resume if we error here?
+
+                let mut inv: MutinyInvoice = invoice.clone().into();
```
you already did this above on 188

benthecarman replied:
you already did this above on 188

benthecarman commented on mutiny-core/src/fedimint.rs:
```diff
@@ -0,0 +1,271 @@
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq, Clone, Default)]
+pub struct FedimintIndex {
+    pub child_index: u32,
+    pub federation_code: String,
```
You should be able to make this `InviteCode` so you don't need to parse it later

benthecarman replied:
looks like there is a bunch of places where you are using `String` would be nice to use `InviteCode` instead

benthecarman commented on mutiny-core/src/nodemanager.rs:
```diff
@@ -1662,75 +1604,72 @@ impl<S: MutinyStorage> NodeManager<S> {
-            );
-            Err(MutinyError::NotFound)
-        }
+        let nodes = self.nodes.lock().await;
+        let node = match self_node_pubkey {
+            Some(pubkey) => nodes.get(pubkey),
+            None => nodes.iter().next().map(|(_, node)| node),
+        };
+        let node = node.ok_or(MutinyError::NotFound)?;
```
should make this a function, you do this a lot

benthecarman commented on mutiny-core/src/lib.rs:
```diff
@@ -359,6 +418,128 @@ impl<S: MutinyStorage> MutinyWallet<S> {
+                }
+                // If invoice amount is None or balance is not sufficient, fall through to node_manager payment
+            }
+            // If federation client is not found, fall through to node_manager payment
```
should also iterate through the fedimint clients if first fails

benthecarman commented on mutiny-core/src/lib.rs:
```diff
@@ -359,6 +418,128 @@ impl<S: MutinyStorage> MutinyWallet<S> {
+    ///
+    /// This will not include any funds in an unconfirmed lightning channel.
+    pub async fn get_balance(&self) -> Result<MutinyBalance, MutinyError> {
+        Ok(MutinyBalance::new(self.node_manager.get_balance().await?))
```
fedimint balance should be added

TonyGiorgio replied:
No fedimint funds exist at this stage in the commit but yes federation balances have been added. I'd rather not worry about the details of multi federation yet.

benthecarman reviewed (commented):
looking really good! just a couple nits

TonyGiorgio commented on mutiny-core/src/sql/glue.rs:
```diff
@@ -0,0 +1,566 @@
+#![allow(dead_code, unused_variables, unused_imports, unreachable_code)] // TODO remove
```
@elsirion could you check the approach for this `glue.rs` file whenever you have a chance?

The approach I've taken for the fedimint traits is to have a KV table on top of gluesql (w/ indexeddb under the hood) where each fedimint client will have their own row. I'm just doing a dump of the values on `commit` like you are with webimint. I tried a few different approaches but came back to that one being the safest while still adhering to the trait requirements.

TonyGiorgio commented:
I'm thinking about going ahead and implementing `MutinyStorage` for `GlueDB`, but as another underlying `GlueKV` struct that is specifically just for KV data like fedimint and our existing data. Then `GlueKV` can pull in `VSSClient` and handle all of the versioning and existing data that we're storing as a KV anyways (all of it currently).

If I do that, then at that point we can just do a clean migration from IndexedDB to Glue. 

Then at a later point, we can start to add structured SQL queries for data, like payments and NWC things. However, I'm still a bit unsure of how we should best handle the structured SQL to KV after we migrate to that. 

I'm also wondering if we can now get rid of the in memory preload of all of the data. We could do the `GlueKV` `VSS` integration always doing the remote version check for `get` and then retrieve if there's a difference. Would speed up initial load times. 

Lastly, I'm wondering how much farther we should go with this PR before we ship an early alpha of fedimint. I'd like to get VSS integrations in there first but that opens all of these cans of worms. It's definitely possible we go without it and just warn our early alpha users that these funds are subject to complete loss (which is to be expected whether we have VSS or not for this testing period). 

Thoughts @benthecarman ?

benthecarman commented:
If that's the plan I think it'd make sense to merge this before vss. 

I have been thinking about if we can remove the memory preload. I think it'd be possible because we can use async storage pretty much everywhere. If anything still needs it we could just preload that.

TonyGiorgio commented:
I think one of the other last things to do is making sure we can do NWC stuff out of the fedimint. I should probably do that before getting into complicated storage stuff.

benthecarman commented:
Yeah I think we should hold off on touching the rest of storage until we have fedimint working in production

TonyGiorgio commented on mutiny-wasm/src/models.rs:
```diff
@@ -417,17 +418,88 @@ impl MutinyBalance {
+#[derive(Serialize, Deserialize, Clone, Eq, PartialEq)]
+#[wasm_bindgen]
+pub struct FederationBalance {
+    identity: FedimintIdentity,
```
Rename all `Fedimint` references to `Federation` for consistency.

TonyGiorgio commented:
Could use another review. Last thing I need to hook up is the zap stuff, but otherwise I've cleaned this up a ton and handled all previous PR feedback. Tested pretty thoroughly too.

benthecarman commented on mutiny-core/src/federation.rs:
```diff
@@ -0,0 +1,612 @@
+    sync::{atomic::AtomicBool, Arc, RwLock},
+};
+
+const FEDIMINT_CLIENT_NONCE: &[u8] = b"Fedimint Client Salt";
```
is this supposed to be a fixed value? would be nice to have a comment explaining what this does/is for

benthecarman replied:
maybe we should put the name mutiny in it to so we don't clash with other clients? unless we want to?

TonyGiorgio replied:
If someone derrives their `m/1'/X`keys, they could use it in theory with other fedimint clients. So it's a feature that might not ever be necessary but could be a bonus for interoperability. I added a comment about it.

benthecarman commented on mutiny-core/src/federation.rs:
```diff
@@ -0,0 +1,612 @@
+#[derive(Debug, Serialize, Deserialize, Clone, Default)]
+pub struct FederationStorage {
+    pub federations: HashMap<String, FederationIndex>,
+    #[serde(default)]
```
shouldn't need this `#[serde(default)]`, only needed that for the node storage because we had ones without a version.

Might be better to just use `VersionedValue` instead so we don't need to add another case for serialization

TonyGiorgio replied:
Idk about `VersionedValue`. We're only currently doing it for the channel manager, and I think that's mostly so that we can compare it's internal value with the value stored, right? Why don't we use it for channel monitor, nodes, locks, etc? 

Also, I would rather be passing around `FederationStorage` to multiple places instead of passing around and parsing a generic `Value`. We've got to parse it somewhere, and should be using real values. Plus we never know what we're going to add to this later.

TonyGiorgio replied:
Not handling for now, will worry about this as we tighten up more storage stuff.

benthecarman commented on mutiny-core/src/federation.rs:
```diff
@@ -0,0 +1,612 @@
+    stop: Arc<AtomicBool>,
+}
+
+// TODO remove
+#[allow(dead_code)]
```
think you can remove this now

benthecarman commented on mutiny-core/src/federation.rs:
```diff
@@ -0,0 +1,612 @@
+        // a list of components that need to be stopped and whether or not they are stopped
+        let stopped_components = Arc::new(RwLock::new(vec![]));
+
+        let stop = Arc::new(AtomicBool::new(false));
```
shouldn't we be getting this from `lib.rs`?

benthecarman commented on mutiny-core/src/federation.rs:
```diff
@@ -0,0 +1,612 @@
+                let logger = self.logger.clone();
+                let lightning_module = Arc::clone(&lightning_module);
+                async move {
+                    if entry.operation_module_kind() == "ln" {
```
is there a way to get around this magic string, can we match on the type instead?

benthecarman commented on mutiny-core/src/federation.rs:
```diff
@@ -0,0 +1,612 @@
+            .get_first_module::<LightningClientModule>();
+
+        for (key, entry) in operations {
+            if entry.operation_module_kind() == "ln" {
```
same here

benthecarman commented on mutiny-core/src/federation.rs:
```diff
@@ -0,0 +1,612 @@
+        let operations = self
+            .fedimint_client
+            .operation_log()
+            .list_operations(100, None)
```
i think this only looks back at the last hundred, we'll need to add handling if you've done over 100 txs

benthecarman commented on mutiny-core/src/federation.rs:
```diff
@@ -0,0 +1,612 @@
+
+    pub(crate) async fn pay_invoice(
+        &self,
+        invoice: &Bolt11Invoice,
+    ) -> Result<MutinyInvoice, MutinyError> {
+        let lightning_module = self
+            .fedimint_client
+            .get_first_module::<LightningClientModule>();
+        let outgoing_payment = lightning_module
+            .pay_bolt11_invoice(invoice.clone(), ())
```
if we're just immediately cloning then no need to pass in a reference

benthecarman commented on mutiny-core/src/federation.rs:
```diff
@@ -0,0 +1,612 @@
+                            process_pay_state_internal,
+                            invoice.clone(),
+                            true,
+                            30_000,
```
should just pull in `DEFAULT_PAYMENT_TIMEOUT * 1_000` from `node.rs`  instead of this magic value

benthecarman commented on mutiny-core/src/federation.rs:
```diff
@@ -0,0 +1,612 @@
+                            process_pay_state_ln,
+                            invoice.clone(),
+                            false,
+                            30_000,
```
same here

benthecarman commented on mutiny-core/src/federation.rs:
```diff
@@ -0,0 +1,612 @@
+    Ok(secret)
+}
+
+fn network_to_bip44_chain(network: Network) -> u32 {
+    match network {
+        Network::Bitcoin => 0,
+        Network::Testnet => 1,
+        Network::Signet => 1,
+        Network::Regtest => 1,
+    }
+}
```
we are doing this in `onchain.rs`, should move this function to utils and call it in both places

benthecarman commented on mutiny-core/src/federation.rs:
```diff
@@ -0,0 +1,612 @@
+                            process_pay_state_ln,
+                            pay_meta.invoice,
+                            false,
+                            30,
```
this 30 number is here again but not multiplied by 1k, probably a bug? at the very least should be a `const`

benthecarman commented on mutiny-core/src/federation.rs:
```diff
@@ -0,0 +1,612 @@
+            if invoice.payment_hash() == hash {
+                match lightning_module.subscribe_ln_receive(operation_id).await {
+                    Ok(o) => Some(
+                        process_outcome(o, process_receive_state, invoice, true, 30, logger).await,
```
same here

benthecarman commented on mutiny-core/src/federation.rs:
```diff
@@ -0,0 +1,612 @@
+        LightningOperationMetaVariant::Pay(pay_meta) => {
+            match lightning_module.subscribe_ln_pay(operation_id).await {
+                Ok(o) => {
+                    process_outcome(o, process_pay_state_ln, pay_meta.invoice, false, 30, logger)
```
same here

benthecarman commented on mutiny-core/src/federation.rs:
```diff
@@ -0,0 +1,612 @@
+        }
+        LightningOperationMetaVariant::Receive { invoice, .. } => {
+            match lightning_module.subscribe_ln_receive(operation_id).await {
+                Ok(o) => process_outcome(o, process_receive_state, invoice, true, 30, logger).await,
```
same here

benthecarman commented on mutiny-core/src/lib.rs:
```diff
@@ -355,6 +532,201 @@ impl<S: MutinyStorage> MutinyWallet<S> {
+        for federation_id in federation_ids {
+            if let Some(fedimint_client) = self.federations.lock().await.get(&federation_id) {
+                // Check the amount specified in the invoice
+                if let Some(invoice_amount) = inv.amount_milli_satoshis() {
```
should do this check before `list_federations` otherwise we'll have to loop through each time for no reason

benthecarman commented on mutiny-core/src/lib.rs:
```diff
@@ -355,6 +532,201 @@ impl<S: MutinyStorage> MutinyWallet<S> {
+        labels: Vec<String>,
+    ) -> Result<MutinyBip21RawMaterials, MutinyError> {
+        let invoice = if self.config.safe_mode {
+            None
```
we can probably enable fedimint in safe mode? probably a better discussion for the future however

TonyGiorgio replied:
Nah, in case something is going wrong with fedimint, it would be good for now to leave that unaccessible for now. Until we test more at least.

benthecarman commented on mutiny-core/src/lib.rs:
```diff
@@ -630,6 +994,231 @@ impl<S: MutinyStorage> MutinyWallet<S> {
+
+    pub async fn get_federation_balances(&self) -> Result<FederationBalances, MutinyError> {
+        let federation_ids = self.list_federations().await?;
+        let mut balances = Vec::new();
```
nit: can do `let mut balances = Vec::with_capacity(federation_ids.len());`

benthecarman reviewed (commented):
Overall looking good, mostly nits

Would be good to try and document at least all the public functions so it easier to grok as well.

benthecarman commented on mutiny-core/src/nodemanager.rs:
```diff
@@ -1461,7 +1326,7 @@ impl<S: MutinyStorage> NodeManager<S> {
             .sum();
 
-        Ok(MutinyBalance {
+        Ok(OnChainLnBalance {
```
Maybe `NodesBalance`

TonyGiorgio commented on mutiny-core/src/federation.rs:
```diff
@@ -0,0 +1,612 @@
+        &self,
+        invoice: &Bolt11Invoice,
+    ) -> Result<MutinyInvoice, MutinyError> {
+        let lightning_module = self
```
There's a couple ways we can go about this and I'm unsure of how we should do it. 

Unfortunately, we save a ton of things inside of `LdkStorage` and this includes all payment information. Another unfortunate thing is we append the node_id to all of the keys. So, if we want to store and recheck if we've made a payment, there's a few options: 
1. Ignore if we possibly paid it through the self custodial node
2. Pull in `storage` and save it similarly on the fedimint side in the KV store (only checking fedimint)
3. Move payments to GlueSQL for fedimint or also for node payments
4. Do nothing for now

benthecarman replied:
Maybe a combo of 1 and 4 would be good here. Start storing fedimint payments in gluesql and we can plan to move the node ones over in a future update

benthecarman commented:
We can pull out the `Move log and stop outside of nodemanager ` commit separately and just merge that for now if we want, not sure how close we are getting here

benthecarman commented on mutiny-core/src/lib.rs:
```diff
@@ -525,6 +526,31 @@ impl<S: MutinyStorage> MutinyWallet<S> {
             return Err(MutinyError::IncorrectNetwork(inv.network()));
         }
 
+        // Try each federation first
```
this comment needs to be moved down

benthecarman commented on mutiny-core/src/federation.rs:
```diff
@@ -0,0 +1,216 @@
+    }
+}
+
+// A federation private key will be derived from `m/1'/N'/X'`, where:
+// - N is the network type
+// - X is the index of a specific federation.
+//
+// Federation will derive further keys from there.
+fn create_federation_secret(
+    xprivkey: ExtendedPrivKey,
+    federation_index: &FederationIndex,
+    network: Network,
+) -> Result<DerivableSecret, MutinyError> {
+    let context = Secp256k1::new();
+    let xpriv = xprivkey.derive_priv(
+        &context,
+        &DerivationPath::from(vec![
+            ChildNumber::from_hardened_idx(1)?,
+            ChildNumber::from_hardened_idx(coin_type_from_network(network))?,
+            ChildNumber::from_hardened_idx(federation_index.child_index)?,
+        ]),
+    )?;
+    let secret =
+        DerivableSecret::new_root(&xpriv.private_key.secret_bytes(), FEDIMINT_CLIENT_NONCE);
+    Ok(secret)
+}
```
Was thinking about this more, do we need the `federation_index`? Don't see why we need different keys for each federation. I don't think there is a privacy implication since it is all ecash, could be wrong though.

If we got rid of it it'd make things easier where if the user has their seed they don't also need to know the index for each federation.

TonyGiorgio replied:
This is a valid point. I can remove it and just use the master seed but be aware, both fedimint and our LN nodes will be using `m/0'/` as the first part of the prefix. I'm afraid of potential clashes.

TonyGiorgio replied:
These secondary indexes should be hardcoded and stored somewhere.

benthecarman commented on mutiny-core/src/lib.rs:
```diff
@@ -876,6 +893,182 @@ impl<S: MutinyStorage> MutinyWallet<S> {
+    }
+}
+
+async fn create_federation<S: MutinyStorage>(
```
nit: this should be named `create_federations`

benthecarman commented on mutiny-core/src/lib.rs:
```diff
@@ -876,6 +893,182 @@ impl<S: MutinyStorage> MutinyWallet<S> {
+    // so that we can create another federation with the next.
+    // Always get it from our storage, the federation_mutex is
+    // mostly for read only and locking.
+    let mut existing_federations = storage.get_federations()?;
```
should we check that the new federation isn't in our current list. Could see where you archive one and then add it back, we'd have it twice in the list, could just unarchive in that case, if it is active then can throw an error.

benthecarman commented on mutiny-core/src/lib.rs:
```diff
@@ -876,6 +893,182 @@ impl<S: MutinyStorage> MutinyWallet<S> {
+    )
+    .await;
+
+    let new_federation = match new_federation_res {
+        Ok(new_federation) => new_federation,
+        Err(e) => return Err(e),
+    };
```
pretty sure this is the exact same as `new_federation_res?`

benthecarman commented on mutiny-core/src/lib.rs:
```diff
@@ -876,6 +893,182 @@ impl<S: MutinyStorage> MutinyWallet<S> {
+
+    let federation_id = new_federation.fedimint_client.federation_id();
+    federations
+        .clone()
```
dont think you need this clone

benthecarman commented on mutiny-core/src/lib.rs:
```diff
@@ -655,7 +668,28 @@ impl<S: MutinyStorage> MutinyWallet<S> {
+        let mut activities = self.node_manager.get_activity().await?;
+
+        // Get federation activities and merge them with node manager activities
+        let federation_ids = self.list_federations().await?;
+        for fed_id in federation_ids {
+            let federation_activities = self
+                .federations
+                .lock()
+                .await
+                .get(&fed_id)
+                .ok_or(MutinyError::NotFound)?
+                .get_activity()
```
this seems inefficient, we get the ids and individually get each client, can we not do this?

```rust 
let federations = self.federations.lock().await;

for (fed_id, federation) in federations {
	let federation_activities= federation.get_activity().await?;
	activities.extend(federation_activities);
}

```

benthecarman commented on mutiny-core/src/lib.rs:
```diff
@@ -670,7 +704,23 @@ impl<S: MutinyStorage> MutinyWallet<S> {
+
+        // If not found, search in federations
+        // FIXME: this is incredibly inefficient without a KV mapping operationid:hash
+        let federation_ids = self.list_federations().await?;
+        for fed_id in federation_ids {
+            if let Some(fed_client) = self.federations.lock().await.get(&fed_id) {
+                if let Ok(invoice) = fed_client.get_invoice_by_hash(hash).await {
+                    return Ok(invoice);
+                }
+            }
+        }
```
same here

benthecarman commented on mutiny-core/src/sql/glue.rs:
```diff
@@ -0,0 +1,609 @@
+
+    #[tokio::test]
+    async fn create_glue_storage_tests() {
+        // TODO reactivate
+        // create_glue_and_fedimint_storage().await;
+    }
+
+    #[tokio::test]
+    async fn create_glue_storage_value_tests() {
+        // TODO reactivate
+        // create_glue_storage_value().await;
```
can we enable these now?

TonyGiorgio replied:
No, glue can't run on a server. I may need to just delete these.

benthecarman commented on mutiny-core/src/sql/glue.rs:
```diff
@@ -0,0 +1,609 @@
+    }
+}
+
+fn key_id(federation_id: String) -> String {
```
nit could make this a `&str` so you don't need clones

benthecarman reviewed (commented):
Went through commit by commit this time. Mostly nits again, what is left before we can merge this?

TonyGiorgio commented:
I got payment status working though labels don't for some reason. Going to clean up and get that working tomorrow.

TonyGiorgio commented on mutiny-core/src/nostr/nwc.rs:
```diff
@@ -1295,13 +1316,21 @@ mod wasm_test {
+        let result = nwc.handle_nwc_request(event, &mw, &nostr_manager).await;
         assert_eq!(result.unwrap(), None);
-        check_no_pending_invoices(&storage);
+        // FIXME: Not working
```
@benthecarman I couldn't get a couple of these NWC asserts to work and I've tried for quite a bit. I tested budgetting and auto approval pretty thoroughly for the node & fedimint payer, so I don't think I've introduced a bug here. TBH, I'm a bit unsure how how these tests actually work since I don't think these nodes have funds and don't attempt to pay? 

We can either fix in a follow up or if you can investigate this further, go ahead.

TonyGiorgio commented:
This is ready @benthecarman 

I say we handle LNURL pay/withdraw in a follow up along with you fixing a couple of the NWC tests in a follow up too. I don't think they'd require any foundational change, mostly a bit of manual work that's not worth blocking on.

benthecarman commented on mutiny-core/src/lib.rs:
```diff
@@ -93,15 +93,14 @@ use mockall::{automock, predicate::*};
         &self,
         invoice: &Bolt11Invoice,
         amt_sats: Option<u64>,
-        timeout_secs: Option<u64>,
```
since you removed the timeout can rename this to just `pay_invoice`

benthecarman commented on mutiny-core/src/lib.rs:
```diff
@@ -440,6 +439,7 @@ impl<S: MutinyStorage> MutinyWallet<S> {
     pub(crate) async fn start_nostr_wallet_connect(&self, from_node: PublicKey) {
         let nostr = self.nostr.clone();
         let nm = self.node_manager.clone();
```
i think we're only getting the node manager for the logger now, could just replace with self.logger

TonyGiorgio replied:
Glad you caught this. It's more involved because `node` is being retrieved and called upon actually. Not everywhere was it using the `LnPayer` trait to find the status of a payment. But I should be able to get this tonight.

benthecarman reviewed (commented):
only 2 nits

benthecarman commented on mutiny-core/src/lib.rs:
```diff
@@ -653,33 +660,9 @@ impl<S: MutinyStorage> MutinyWallet<S> {
-                    .create_invoice(amount, labels.clone())
-                    .await?;
-                Some(inv.bolt11.ok_or(MutinyError::WalletOperationFailed)?)
+            match self.create_lightning_invoice(amount, labels.clone()).await {
+                Ok(invoice) => Some(invoice), // Assuming create_lightning_invoice returns a Result
+                Err(_) => None, // You can handle the error more specifically if needed
```
We were returning an error here before, should probably keep that functionality at least for now.
otherwise this is the exact same as `ok()`

TonyGiorgio replied:
We didn't return error here, when we create bip21, we return the address that's created but without the invoice, and display a warning.

TonyGiorgio replied:
This is what it was before this PR: 

```
        } else {
            match self.create_lightning_invoice(amount, labels.clone()).await {
                Ok(invoice) => Some(invoice),
                Err(_) => None,
            }
        };
```

I was incorrect in making it return an error when I introduced fedimint.

benthecarman replied:
Okay, we can at least clean this up a bit with `ok()` if you want

TonyGiorgio replied:
Doing `ok().and_then()` here now.

TonyGiorgio commented:
@benthecarman please review the last 3 commits.
I've done the following: 
- Removed more references to `node_manager` and `node` inside of `nwc`
- Added another method to the trait in order to handle invoice creation (for one time nwc)
- I've renamed `LnPayer` to `InvoiceHandler` since it now generates, checks, and pays invoices

benthecarman commented on mutiny-core/src/lib.rs:
```diff
@@ -689,12 +672,40 @@ impl<S: MutinyStorage> MutinyWallet<S> {
+
+        // Fallback to node_manager invoice creation if no federation invoice created
+        self.node_manager
+            .create_invoice(amount, labels.clone())
```
nit: unneeded clone

benthecarman reviewed (commented):
one tiny nit, otherwise lgtm.

You can just take my commit and merge it into your last one if you want, would make the commits a little cleaner.

TonyGiorgio commented:
> one tiny nit, otherwise lgtm.
> 
> You can just take my commit and merge it into your last one if you want, would make the commits a little cleaner.

Done

benthecarman commented:
:ship: it 🚀
